### DISCUSSION TITLE
Bipartite graph sampling / K-hop sampling

### DISCUSSION THREAD

**Discussion by alexeypustynnikov** [pyg-team/pytorch_geometric#2898] [Q&A] [answered]
Hello @rusty1s ! Thank you for your project!

I have some difficulties with implementing two types of sampling. My data consists of 2 graph structures: ordinary graph **G** and bipartite graph **B**. I need to sample from both and apply some aggregation (both pooling and message passing, it depends on approach) to the sampled subgraph.

Can you help me to find the right way to perform these actions?

1. **How can one effectively sample everything in k-hop neighborhood from G? Maybe for several k simultaneously.** 

     Recently I’ve found ShaDowKHopSampler (https://pytorch-geometric.readthedocs.io/en/latest/modules/data.html#torch_geometric.data.ShaDowKHopSampler), but currently it hasn’t got a direct parameter for sampling all neighbors. I found a workaround to this: num_neighbors should equal to max degree in the graph, but this approach seems a bit strange. 

      Maybe there is some way to sample k-hop neighbors by NeighborSampler?

2. **How can one effectively sample from bipartite graph?** 

    Namely, I need to sample from one side (source) of bipartite graph **B** keeping all edges associated with sampled nodes. 
    
    My only idea was to manually iteratively create BipartiteData(Data) (https://pytorch-geometric.readthedocs.io/en/latest/notes/batching.html) using part of edge list corresponding to selected source nodes. It works quite slowly, maybe because of calling BipartiteData object each time.

Thanks in advance!

**Comment by rusty1s** [comment, marked as answer]
We are actively working on sampling support for bipartite graphs/heterogeneous graphs, so please stay tuned :)

1. For sampling the complete neighborhood, you can either use `NeighborSampler` with `sizes=[-1] * k` or use `torch_geometric.utils.k_hop_subgraph`.

2. We do not have a dedicated implementation for that yet. If possible, I suggest to create one big graph (containing both `G` and `B`), and sample from their, e.g, by stacking `G`, and `B` into a single adjacency matrix, as we do [here](https://github.com/snap-stanford/ogb/blob/master/examples/nodeproppred/mag/sampler.py#L57-L68).

**Reply by alexeypustynnikov** [reply]
@rusty1s Thanks for the answer! It helps.
